addons/hw_drivers/websocket_client.py

```diff
@@ -115,3 +54,90 @@ def run(self):
+        """Confirm the operation's completion by sending
+        a response back to the Odoo server.
+
+        :param params: The parameters to send back to the server
+        """
+        params['iot_mac'] = helpers.get_mac_address()
+        server_url = helpers.get_odoo_server_url() + "/iot/device/status"
+        try:
+            response = requests.post(
+                server_url,
+                data=json.dumps({'params': params}),
+                headers={'Content-type': 'application/json'},
+            )
+            response.raise_for_status()
+        except requests.exceptions.RequestException:
+            _logger.exception('Could not reach confirmation status URL: %s', server_url)
+
+    def on_open(self, ws):
+        """When the client is set up, this function sends a message
+        to subscribe to the iot websocket channel.
+
+        :param ws: The websocket client
+        """
+        ws.send(
+            json.dumps({
+                'event_name': 'subscribe',
+                'data': {
+                    'channels': [self.iot_channel],
+                    'last': 0,
+                    'mac_address': helpers.get_mac_address()
+                }
+            })
+        )
+
+    def on_message(self, _ws, messages):
+        """Synchronously handle messages received by the websocket.
+
+        :param _ws: The websocket client
+        :param messages: The messages received by the websocket
+        """
+        messages = json.loads(messages)
+        _logger.debug("websocket received a message: %s", pprint.pformat(messages))
+        iot_mac = helpers.get_mac_address()
+        for message in messages:
+            msg_type = message['message']['type']
+            msg_payload = message['message']['payload']
+
+            if msg_type == 'operation_confirmation': # ignore confirmation messages
+                return
+            if msg_type != 'iot_action':
+                _logger.warning("Message type not supported: %s", msg_type)
+                return
+
+            if msg_payload['iot_box_identifier'] != iot_mac: # likely intended as IoT share the same channel
+                _logger.debug("Message ignored due to different iot mac: %s", iot_mac)
+                return
+
+            device_identifier = msg_payload['device_identifier']
+            if device_identifier in main.iot_devices:
+                start_operation_time = time.perf_counter()
+                _logger.debug("device '%s' action started with: %s", device_identifier, pprint.pformat(msg_payload))
+                response = {
+                    'operation_id': msg_payload['operation_id'],
+                    'device_identifier': device_identifier,
+                    'status': 'success',
+                }
+                try:
+                    response.update(main.iot_devices[device_identifier].action(msg_payload))
+                except Exception:
+                    response['status'] = 'error'
```

Comment:
we might find another way of catching errors raised by device actions (currently, no action explicitly raises an error)

Reply:
( at least log the exception if you plan to not re-raise it :(